src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs

```diff
@@ -132,99 +137,128 @@ public virtual byte[] GetValue()
             {
                 Debug.Assert(controls[i] != null);
-                byte[] value = controls[i].GetValue();
+                byte[] value = controls[i]._directoryControlValue ?? Array.Empty<byte>();
+                Span<byte> asnSpan = value;
+                bool asnReadSuccessful;
+
                 if (controls[i].Type == "1.2.840.113556.1.4.319")
                 {
-                    // The control is a PageControl.
-                    object[] result = BerConverter.Decode("{iO}", value);
-                    Debug.Assert((result != null) && (result.Length == 2));
+                    // The control is a PageControl, as described in RFC 2696.
+                    byte[] cookie = Array.Empty<byte>();
+
+                    AsnDecoder.ReadSequence(asnSpan, AsnEncodingRules.BER, out int sequenceContentOffset, out int sequenceContentLength, out _);
+                    Debug.Assert(sequenceContentLength > 0);
 
-                    int size = (int)result[0];
-                    // user expects cookie with length 0 as paged search is done.
-                    byte[] cookie = (byte[])result[1] ?? Array.Empty<byte>();
+                    asnReadSuccessful = AsnDecoder.TryReadInt32(asnSpan.Slice(sequenceContentOffset), AsnEncodingRules.BER, out int size, out int bytesConsumed);
+                    Debug.Assert(asnReadSuccessful);
+                    asnSpan = asnSpan.Slice(sequenceContentOffset + bytesConsumed);
 
-                    PageResultResponseControl pageControl = new PageResultResponseControl(size, cookie, controls[i].IsCritical, controls[i].GetValue());
+                    asnReadSuccessful = AsnDecoder.TryReadEncodedValue(asnSpan, AsnEncodingRules.BER, out _, out int _, out int cookieLength, out _);
+                    if (asnReadSuccessful)
+                    {
+                        cookie = new byte[cookieLength];
+
+                        // user expects cookie with length 0 as paged search is done.
+                        asnReadSuccessful = AsnDecoder.TryReadOctetString(asnSpan, cookie, AsnEncodingRules.BER, out _, out _);
+                        Debug.Assert(asnReadSuccessful);
+                    }
+
+                    PageResultResponseControl pageControl = new PageResultResponseControl(size, cookie, controls[i].IsCritical, value);
                     controls[i] = pageControl;
                 }
                 else if (controls[i].Type == "1.2.840.113556.1.4.1504")
                 {
-                    // The control is an AsqControl.
-                    object[] o = BerConverter.Decode("{e}", value);
-                    Debug.Assert((o != null) && (o.Length == 1));
+                    // The control is an AsqControl, as described in MS-ADTS section 3.1.1.3.4.1.18
+                    AsnDecoder.ReadSequence(asnSpan, AsnEncodingRules.BER, out int sequenceContentOffset, out int sequenceContentLength, out _);
+                    Debug.Assert(sequenceContentLength > 0);
```

Comment:
The same feedback from the previous section will apply to all of these object reads.